src/workerd/io/worker.c++

```diff
@@ -1971,6 +1971,10 @@ kj::Maybe<kj::Own<api::ExportedHandler>> Worker::Lock::getExportedHandler(
 
   kj::StringPtr n = name.orDefault("default"_kj);
   KJ_IF_SOME(h, worker.impl->namedHandlers.find(n)) {
+    jsg::Lock& js = *this;
+    if (FeatureFlags::get(js).getCtxPropsToDefault()) {
+      h.ctx = jsg::alloc<api::ExecutionContext>(js, props.toJs(js));
```

Comment:
This is overwriting the `ctx` on the singleton `ExecutionContext` which is shared by all events. This seems unsafe -- a race condition between concurrent events could cause one event to receive the other one's props.

Reply:
Where is the appropriate place to do this assignment?

Reply:
Oops, I meant singleton `ExportedHandler`, not `ExecutionContext`. (You are creating a new `ExecutionContext`, but attaching it to the shared `ExportedHandler`, possibly overwriting a previous value that is still in use.)

We need to refactor the code to create a new `ExportedHandler` instance for each invocation, like we do for `WorkerEntrypoint` entrypoints.

Reply:
would it be as simple as copying h and then returning an owned version of that?

Reply:
`ExportedHandler` currently doesn't have a simple way to copy it. You may need to add a `clone()` method which calls each member's `addRef()` method (`jsg::Function`, `jsg::Value`, `jsg::Ref`, and `jsg::SelfRef` should all have versions of this method).